# HW 3
2020 April 7

Price   $: 7/12

 - Price   $, wait: 3/7

 - Price   $, not wait: 4/7

Price  $$: 2/12

 - Price  $$, wait: 2/2

 - Price  $$, not wait: 0/2

Price $$$: 3/12

 - Price $$$, wait: 1/3

 - Price $$$, not wait: 2/3

$ Remainder(price) = -\frac{7}{12}[\frac{3}{7}\lg(\frac{3}{7}) + \frac{4}{7}\lg(\frac{4}{7})] - \frac{2}{12}[1\lg(1) + 0\lg(0)] - \frac{3}{12}[\frac{1}{3}\lg(\frac{1}{3}) + \frac{2}{3}\lg(\frac{2}{3}]$
$ = 0.804 $ 

$ Gain(price) = 1 - 0.804 = 0.196 $

This is not as big an information gain as it is with Patrons


## 2.
It is possible to simplify the network a little using the structure of the XOR gate.

With the following network, only exclusive or would result in an output of 1. There is one hidden node that has a threshold of 1.5 to be activated and its output is multiplied by -2. This acts as a NOT and prevents inputs of 1 and 1 from giving an output of 2.

![alt text](xor_gate.png)



## 3. 

In [0]:
import numpy as np
import pandas as pd
import keras

from keras.datasets import boston_housing

(train_data, train_targets), (test_data, test_targets) = boston_housing.load_data()

### a. Data Structure Dimensions

In [32]:
print("Training data shape: ", train_data.shape)
print("Testing data shape: ", test_data.shape)

Training data shape:  (404, 13)
Testing data shape:  (102, 13)


In [35]:
training_frame= pd.DataFrame(data = train_data)
training_framep

0     1      2    3       4   ...    8      9     10      11     12
0    1.23247   0.0   8.14  0.0  0.5380  ...   4.0  307.0  21.0  396.90  18.72
1    0.02177  82.5   2.03  0.0  0.4150  ...   2.0  348.0  14.7  395.38   3.11
2    4.89822   0.0  18.10  0.0  0.6310  ...  24.0  666.0  20.2  375.52   3.26
3    0.03961   0.0   5.19  0.0  0.5150  ...   5.0  224.0  20.2  396.90   8.01
4    3.69311   0.0  18.10  0.0  0.7130  ...  24.0  666.0  20.2  391.43  14.65
..       ...   ...    ...  ...     ...  ...   ...    ...   ...     ...    ...
399  0.21977   0.0   6.91  0.0  0.4480  ...   3.0  233.0  17.9  396.90  16.20
400  0.16211  20.0   6.96  0.0  0.4640  ...   3.0  223.0  18.6  396.90   6.59
401  0.03466  35.0   6.06  0.0  0.4379  ...   1.0  304.0  16.9  362.25   7.83
402  2.14918   0.0  19.58  0.0  0.8710  ...   5.0  403.0  14.7  261.95  15.79
403  0.01439  60.0   2.93  0.0  0.4010  ...   1.0  265.0  15.6  376.70   4.38

[404 rows x 13 columns]

### b. Construct a testing set, training set, and validation set for the data

In [33]:
training_frame= pd.DataFrame(data = train_data)
training_frame = training_frame.reindex(
    np.random.permutation(training_frame.index)
)

print("Training frame size: ", training_frame.shape)

Training frame size:  (404, 13)


There are 404 rows in the dataset. The training set will be made of the first 304. The validation set will be the remaining 100. 

In [0]:
training_data = training_frame.head(304)
validation_data = training_frame.tail(100)
testing_data= pd.DataFrame(data = test_data)
testing_data = testing_data.reindex(
    np.random.permutation(testing_data.index)
)

### c. Create one new synthetic feature that could be useful in this domain

I made a synthetic feature that is the property tax per $1000 dollars divided by the crime rate. This will show how the property taxes are affected by crime rate and whether that relationship can predict prices.

In [40]:
training_data[13] = training_data[9]/training_data[0]
training_data

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


0     1      2    3      4   ...     9     10      11     12           13
28    0.67191   0.0   8.14  0.0  0.538  ...  307.0  21.0  376.88  14.81   456.906431
277   0.19073  22.0   5.86  0.0  0.431  ...  330.0  19.1  393.74   6.56  1730.194516
47    1.00245   0.0   8.14  0.0  0.538  ...  307.0  21.0  380.23  11.98   306.249688
89    7.05042   0.0  18.10  0.0  0.614  ...  666.0  20.2    2.52  23.29    94.462458
197  23.64820   0.0  18.10  0.0  0.671  ...  666.0  20.2  396.90  23.69    28.162820
..        ...   ...    ...  ...    ...  ...    ...   ...     ...    ...          ...
77    0.88125   0.0  21.89  0.0  0.624  ...  437.0  21.2  396.90  18.34   495.886525
44   13.35980   0.0  18.10  0.0  0.693  ...  666.0  20.2  396.90  16.35    49.851046
128   0.78570  20.0   3.97  0.0  0.647  ...  264.0  13.0  384.07  14.79   336.006109
131   0.14476   0.0  10.01  0.0  0.547  ...  432.0  17.8  391.50  13.61  2984.249793
326   0.62976   0.0   8.14  0.0  0.538  ...  307.0  21.0  396.90   8.26   487.487297

[304 rows x 14 columns]